src/tools/miri/tests/pass/dyn-upcast.rs

```diff
@@ -430,3 +431,53 @@ fn replace_vptr() {
     let s = S(42);
     invoke_outer(&s);
 }
+
+fn drop_principal() {
```

Comment:
The equivalent code with `transmute` causes [Miri to report UB](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=b15339c843fb98b4e2fcfa6bed9317d2). So something is very odd here that makes this pass.

Is it maybe that these get translated to actual casts, and they get a new vtable? But `Box` can't be in a cast so I don't think so.

Anyway if you think the `transmute` version should be sound then please also add that as a Miri test.

Reply:
> Anyway if you think the `transmute` version should be sound then please also add that as a Miri test.

I don’t see any reason to commit to that at this time.

Reply:
I think we should commit the version regardless, to make it intentional if we change it later :thinking:

Reply:
> I don’t see any reason to commit to that at this time.

After playing around with this more, I agree. I realized that there are already trait-changing casts which are NOPs at runtime, namely when the supertrait vtable is a prefix of the subtrait vtable. We compile these casts to a NOP but we don't want to allow people to do this as a transmute; the fact that this is a NOP is an implementation detail. The same applies to this case: semantically, `Debug + Send` to `Send` is still a meaningful cast, even if we compile it to a NOP.

We do currently make `Debug + Send` to `Debug` a non-meaningful cast; vtables in Miri only keep track of the principal trait. Maybe that is something we should reconsider, to be fully consistent.